### PR TITLE
325 layout

### LINKED ISSUE #325: Don't Use Fixed Port Positions During Layout Unless Necessary
The layout code currently uses the FIXED_POS option as a port constraint. This is needed to properly layout feature groups which have children. However, in cases where such features groups are not present, the FREE or FIXED_SIDE port constraint could be used.

Suggest used FIXED_SIDE in that case and to add a preference that will allow the user of FREE. Experiments have shown that FREE produces better results than the current solution but could result in output features being on the left of the container.

### PR DESCRIPTION
**Pull request by philip-alldredge**
Fixes #325

### CODE REVIEW

File: org.osate.ge.doc/docs/user_guide.md [hunk 1]
```diff
@@ -281,9 +284,11 @@ When an element has been added to the diagram and needs to be positioned or size
 
 Table: Incremental Layout Modes {#tbl:incremental_layout_modes}
 
-### Known Layout Issues and Limitations
+#### Layout Feature Based on Direction
+When the *Layout Features Based on Direction* preference is enabled, the graphical editor will position ports based on the feature's direction. Input features will be placed on the left side and output features will be placed on the right side. If the preference is disabled, the layout algorithm will decide the side on which to place the feature.
 
-* Connections between multiple levels of the diagram hierarchy are not routed out. Such connections usually represent property values such as bindings. Such connections must be routed manually.
+### Known Layout Issues and Limitations
+* Connections between multiple levels of the diagram hierarchy are not routed. Such connections usually represent property values such as bindings. Such connections must be routed manually.
```
Comment (RyanMcilnay) [inline org.osate.ge.doc/docs/user_guide.md]: "Such connections usually represent property values such as bindings. Such connections must be routed manually."

I think combining these sentences would be better. Just a thought.

Possible changes:

    "Such connections must be routed manually and usually represent property values such as bindings."
or

    "Such connections usually represent property values such as bindings and must be routed manually."

File: org.osate.ge.doc/docs/user_guide.md [hunk 2]
```diff
@@ -262,16 +262,19 @@ Layout menu items are available from the *Edit->Layout* menu or by right clickin
...
 Table: Layout Menu Items {#tbl:layout_menu_items}
 
-### Incremental Layout
-When an element has been added to the diagram and needs to be positioned or sized, an incremental layout is performed. The behavior of the incremental layout can be configured by selecting a mode from the preferences dialog.
+### Layout Preferences
+The preferences dialog contains preferences which affect the layout behavior of the graphical editor. To edit layout preferences:
 
 1. Select *Window->Preferences*
 
 2. Select to *OSATE Preferences->Diagram*
```
Comment (RyanMcilnay) [inline org.osate.ge.doc/docs/user_guide.md]: On bullet two, Diagram should be plural.

**Comment by philip-alldredge** [comment]
@RyanMcilnay I have pushed updated to the userguide

File: org.osate.ge/src/org/osate/ge/internal/diagram/runtime/layout/ElkGraphBuilder.java [hunk 3]
```diff
@@ -235,12 +308,10 @@ private void createAndPositionPorts(final ElkNode parentNode, final List<Diagram
 	 * @param dockedElement
 	 * @param portBorderOffset additional offset from the border. Used to set the ELK port border offset for nested ports.
 	 * @param mapping
-	 * @param diagramElementToGraphicPortMap
 	 * @return
 	 */
 	private ElkPort createPort(final ElkNode parent, final PortSide side, final DiagramElement dockedElement,
-			final double portBorderOffset, final LayoutMapping mapping,
-			final Map<DiagramElement, ElkPort> diagramElementToGraphicPortMap) {
+			final double portBorderOffset, final LayoutMapping mapping, final boolean parentHasNestedPorts) {
```
Comment (RyanMcilnay) [inline org.osate.ge/src/org/osate/ge/internal/diagram/runtime/layout/ElkGraphBuilder.java]: parentHasNestedPort not being used in createPort()

**Pull request merged**